### PR TITLE
Add consts to Params for individual networks

### PR DESCRIPTION
**Pull request by tcharding**
Add consts to the `Params` type for the individual networks.

### CODE REVIEW

**Comment by tcharding** [comment]
If this fly's we should do the same for `blockdata::constants::genesis_block`.

File: bitcoin/src/consensus/params.rs [hunk 1]
```diff
@@ -49,7 +49,80 @@ pub struct Params {
...
+            pow_limit: Target::MAX_ATTAINABLE_REGTEST,
+            pow_target_spacing: 10 * 60,            // 10 minutes.
+            pow_target_timespan: 14 * 24 * 60 * 60, // 2 weeks.
+            allow_min_difficulty_blocks: true,
+            no_pow_retargeting: true,
+        }
+    }
+
+    /// Creates parameters set for the given network.    /// Creates parameters set for the given network.
+    #[deprecated(since = "TBD", note = "use mainnet, testnet etc. instead")]
```
Comment (Kixunil) [inline bitcoin/src/consensus/params.rs:125]: I don't think we should jump on this just yet.

File: bitcoin/src/consensus/params.rs [hunk 2]
```diff
@@ -49,7 +49,80 @@ pub struct Params {
 }
 
 impl Params {
-    /// Creates parameters set for the given network.
+    /// Creates mainnet parameters.
+    pub const fn mainnet() -> Self {
```
Comment (Kixunil) [inline bitcoin/src/consensus/params.rs:53]: I guess constants might be better?

**Comment by tcharding** [comment]
I'm not sure if this PR adds any value if we are not deprecating `Params::new()`?

**Comment by apoelstra** [comment]
The constants would be nice to have, but yeah, the value is small.

**Comment by tcharding** [comment]
I'll leave this open then, looks right to go.

File: bitcoin/src/consensus/params.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/consensus/params.rs]: Deprecation has been removed.

File: bitcoin/src/consensus/params.rs (hunk 2 above)
Comment (tcharding) [reply bitcoin/src/consensus/params.rs]: Now uses constants.

**Comment by tcharding** [comment]
Re-based on master, I'm not exactly sure what the state of CI is, but this had no acks so seems ok to rebase.

**Comment by apoelstra** [comment]
I think CI was fixed by #2512 (there was some clippy lint related to `Default`) and should be "permanently" fixed by #2489.

**Review by apoelstra — APPROVED** [review]
ACK 3a56ecc677eb4ead19f00db0a3a728366e51a169

File: bitcoin/src/consensus/params.rs [hunk 3]
```diff
@@ -121,6 +136,7 @@ impl From<Network> for Params {
     fn from(value: Network) -> Self { Self::new(value) }
 }
 
+
```
Comment (Kixunil) [inline bitcoin/src/consensus/params.rs:139]: Seems unneeded.

**Review by Kixunil — APPROVED** [review]
ACK 3a56ecc677eb4ead19f00db0a3a728366e51a169

File: bitcoin/src/consensus/params.rs (hunk 3 above)
Comment (tcharding) [reply bitcoin/src/consensus/params.rs:139]: Can we have a policy that any comment thread that starts with a "nit: ..." message can be resolved by the merger? (Slightly off topic, I just thought of it yesterday and here seems like a good place to mention it.(

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: That sounds fine to me, though should it be "resolved" if the nit isn't actually addressed?

Comment (tcharding) [reply bitcoin/src/consensus/params.rs:139]: I thought "nit" meant "fix if you  are force pushing anyways but if this has two acks then do it later or not at all". I didn't want to click resolve though in case at some later time I had to force push, so I can be reminded of the open nits still - although I usually just fix nits anyways, for time sensitive PRs this matters.

Comment (tcharding) [reply bitcoin/src/consensus/params.rs:139]: Another option would be that nits can be handled by the OP using some other method (personal todo list) and marking resolved for them just means "I saw and read this, thanks"

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: Personally I'm leaning toward "we shouldn't require conversation resolution". I feel like if somebody opens a nit, the nit doesn't get fixed, and we merge the PR anyway, that's fine....but then the PR should show that the nit was raised and was never addressed. If you click "resolve" then it gets hidden and it becomes super annoying to try to find the nit.

Our "one ACK required" setting already prevents merging if anybody has left a review where they marked "changes requested". Maybe we should just be more disciplined about using "changes requested" for things that we actually want to be addressed.

Comment (Kixunil) [reply bitcoin/src/consensus/params.rs:139]: You seem to be overcomplicating things. :) If I ACKed anyway it implies this is fine and just found something that could be improved if wanted. In this case simply replying "I don't want to fix it" and clicking resolve is enough.

IOW I hereby give anyone permission to resolve my comments within the review in which I ACKed a PR with either of these actions:

* PR author fixes the issue
* PR author opens a new issue for it (or anyone else who believes the PR is urgent)
* anyone writes a "not worth bothering" response

> Personally I'm leaning toward "we shouldn't require conversation resolution".

I think ideally the rule would be "resolve all conversation except if the reviewer ACKed". AFAIK this is not automatable with GitHub.

> Maybe we should just be more disciplined about using "changes requested"

The issue is that sometimes I'm not confident that something is wrong and want a second opinion. In those cases I don't click changes requested but I want my comment to be addressed - I don't want it to be missed.

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: > In this case simply replying "I don't want to fix it" and clicking resolve is enough.

In this case that's fine, it's just a whitespace that has no effect on the code. But what often happens is that a "nit" turns out, years later, to be the only conversation we had about a design change that seemed trivial at the time but later turned out to be consequential.

And then when I go through the PR looking for whether this issue was brought up, instead of having a clear "nit: this makes X do Y when Z happens, but it's probably fine" with no response, I have 2 dozen "conversation resolved" that I have to manually click on to open up, and probably every single one is some irrelevant thing about a version of the PR that wasn't even merged. But maaybe it turns out that one is relevant.

Ok, if we have a policy that you have to explicitly say "not worth bothering", that will save me the time of reverse-engineering whether an apparently-relevant conversation actually even applies to the version of the code that was merged. But it's still a very bad user experience.

TLDR My issue is that the "resolve" button basically deletes the conversation for practical purposes.

Comment (Kixunil) [reply bitcoin/src/consensus/params.rs:139]: Hmm, yeah, that's annoying. But is it really worse than forgetting to address a review comment? I suspect digging into history is less frequent than forgetting.

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: As a recent example, from the last 5 minutes, see https://github.com/rust-bitcoin/rust-bitcoin/pull/2530#pullrequestreview-1922126999 -- this turned up an issue with a 2-year-old PR with 250+ comments. There are plenty of still-visible comments. The issue under question is not one of them. This is useful information.

Comment (Kixunil) [reply bitcoin/src/consensus/params.rs:139]: Well that example has a bunch of junk there as well that would get hidden if we clicked resolved.

Maybe we could require that whoever merges and clicks "Resolve" needs to record that they did so with justification in a comment. The justification can be as simple as a new issue number. Formatting changes are exempt from this rule.

This would improve the situation since the unresolved things would be grouped and easily found.

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: >The justification can be as simple as a new issue number.

Hmmm, *this* I can get behind. You're allowed to click "resolve" on unresolved stuff but you need to file an issue.

>Formatting changes are exempt from this rule.

For sure.

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: It's still not perfect; when I use `git blame` I wind up on the PR rather than having direct access to filed issues. But I think when you open an issue from a comment, it shows up in the PR as "apoelstra referenced this in issue #whatever" and that notification does not get hidden.

Comment (Kixunil) [reply bitcoin/src/consensus/params.rs:139]: I meant writing justification in a separate comment in the PR - not a discussion. So you'd always see it as the last comment. But yes "referenced by" is also helpful.

Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: Ah, gotcha! Ok, ack, let's try that.

**Review by sanket1729 — APPROVED** [review]
ACK 3a56ecc677eb4ead19f00db0a3a728366e51a169

File: bitcoin/src/consensus/params.rs (hunk 3 above)
Comment (apoelstra) [reply bitcoin/src/consensus/params.rs:139]: Ok, I think we've resolved this side-discussion about resolving discussions.

Meanwhile the original line of code under consideration was a style nit :P. So @Kixunil can we hit "resolve" on this and merge?

Comment (Kixunil) [reply bitcoin/src/consensus/params.rs:139]: Of course.
